homeassistant/components/device_tracker/geofency.py

```diff
@@ -4,129 +4,32 @@
-})
+DEPENDENCIES = ['geofency']
 
 
 def setup_scanner(hass, config, see, discovery_info=None):
-    """Set up an endpoint for the Geofency application."""
-    mobile_beacons = config.get(CONF_MOBILE_BEACONS) or []
-
-    hass.http.register_view(GeofencyView(see, mobile_beacons))
-
-    return True
-
-
-class GeofencyView(HomeAssistantView):
-    """View to handle Geofency requests."""
-
-    url = URL
-    name = 'api:geofency'
-
-    def __init__(self, see, mobile_beacons):
-        """Initialize Geofency url endpoints."""
-        self.see = see
-        self.mobile_beacons = [slugify(beacon) for beacon in mobile_beacons]
-
-    async def post(self, request):
-        """Handle Geofency requests."""
-        data = await request.post()
-        hass = request.app['hass']
-
-        data = self._validate_data(data)
-        if not data:
-            return ("Invalid data", HTTP_UNPROCESSABLE_ENTITY)
-
-        if self._is_mobile_beacon(data):
-            return await self._set_location(hass, data, None)
-        if data['entry'] == LOCATION_ENTRY:
-            location_name = data['name']
-        else:
-            location_name = STATE_NOT_HOME
-            if ATTR_CURRENT_LATITUDE in data:
-                data[ATTR_LATITUDE] = data[ATTR_CURRENT_LATITUDE]
-                data[ATTR_LONGITUDE] = data[ATTR_CURRENT_LONGITUDE]
-
-        return await self._set_location(hass, data, location_name)
-
-    @staticmethod
-    def _validate_data(data):
-        """Validate POST payload."""
-        data = data.copy()
-
-        required_attributes = ['address', 'device', 'entry',
-                               'latitude', 'longitude', 'name']
-
-        valid = True
-        for attribute in required_attributes:
-            if attribute not in data:
-                valid = False
-                _LOGGER.error("'%s' not specified in message", attribute)
-
-        if not valid:
-            return False
-
-        data['address'] = data['address'].replace('\n', ' ')
-        data['device'] = slugify(data['device'])
-        data['name'] = slugify(data['name'])
-
-        gps_attributes = [ATTR_LATITUDE, ATTR_LONGITUDE,
-                          ATTR_CURRENT_LATITUDE, ATTR_CURRENT_LONGITUDE]
-
-        for attribute in gps_attributes:
-            if attribute in data:
-                data[attribute] = float(data[attribute])
-
-        return data
-
-    def _is_mobile_beacon(self, data):
-        """Check if we have a mobile beacon."""
-        return 'beaconUUID' in data and data['name'] in self.mobile_beacons
-
-    @staticmethod
-    def _device_name(data):
-        """Return name of device tracker."""
-        if 'beaconUUID' in data:
-            return "{}_{}".format(BEACON_DEV_PREFIX, data['name'])
-        return data['device']
-
-    async def _set_location(self, hass, data, location_name):
+    """Set up the Geofency device tracker."""
+    @callback
+    def _set_location(device, gps, location_name, attributes):
```

Comment:
Make this a coroutine.

Reply:
Doesn't seem to work when I do it that way. Getting the following error: `/usr/lib/python3.6/asyncio/events.py:145: RuntimeWarning: coroutine '_set_location' was never awaited`